### DISCUSSION TITLE
Images from Azure Blob Storage

### DISCUSSION THREAD

**Discussion by Craigoman79** [QuestPDF/QuestPDF#889] [Q&A]
I am generating a rather large at times PDF which includes a number of dynamic images stored in Azure Blob Storage.
My current approach is to load them up front and reference them in my document generation. 

`var imagesData = new Dictionary<string, byte[]>();
        var imageMessages = TotalMessages from my model;

        int totalImages = imageMessages.Count;
        int currentImageIndex = 0;

        foreach (var message in imageMessages)
        {
            string imageKey = $"{fileId}/Input/{message.MediaUrl}";

            if (!imagesData.ContainsKey(imageKey))
            {
                byte[] imageData = await DownloadImageAsByteArrayAsync(imageKey);
                if (imageData != null)
                {
                    imagesData[imageKey] = imageData;
                };
            }
        }`

and then reference like this.
`string imageKey = $"{fileId}/Input/{message.MediaUrl}";
byte[] imageData = imagesData.ContainsKey(imageKey) ? imagesData[imageKey] : null;
if (imageData != null)
{
    innerColumn.Item().Width(200).Image(imageData).FitArea().WithCompressionQuality(ImageCompressionQuality.Medium);
}
else
{
    _logger.LogError($"Unable to load image from path: {imageKey}");
    innerColumn.Item().Text("Image could not be loaded.");
}`

This works fine locally but not when published to my Azure App service - I run out of memory.

What's the best approach here?
1. Can I reference images directly in my blob storage?
2. Should I save them to a tmp folder and then dispose of them afterwards
3. Do they need to exist at time of calling the .GeneratePdf(localFilePath);

I need my method to support long complex pdf documents with dynamic images.
Please help.

**Comment by Craigoman79** [comment]
any insights here?

**Comment by MarcinZiabek** [comment]
Could you provide more details?

1) How much RAM is taken by document generation? How much is available on the hosted environment?
2) How long are your documents?
3) How many images are within a single document?
4) What is the original size and format of an image?

**Comment by Craigoman79** [comment]
My current test case is outputting 45 A4 pages.
68 images of around 200k each.

My end goal is a much bigger document I would presume 2000+ pages with around 4000 images of between 200-300k.

I have modified my code to rather use tmp storage and using SkiaSharp to compress the images by 50% as I am forcing smaller images on the doc. This has allowed me to execute the small test case but the generation is rather slow.

I can chunk the export and merge afterwards if that would be a better approach.
Is there any option to link an image to an existing Azure blob storage rather than a byte array or local file?

**Reply by MarcinZiabek** [reply]
Newer QuestPDF releases automatically resize and compress images to fit the target DPI and image quality best. However, this process requires the image to be in memory during the entire document generation process. You can disable this feature by using `UseOriginalImage` after putting the image in FluentAPI.

70 images of the provided size is nothing. I don't expect any issues at such a small test scale. 4000 images is a different story.

I assume you are using the latest QuestPDF release? Just in case, could you please also test with `2023.12.X` to check if there are any improvements/regressions?

How much RAM is taken during the generation process? 
How much time does it take? 
How much RAM is available on local and production environments?

**Comment by Craigoman79** [comment]
For my sample of 45 pages. these are the results of the overall process using both versions. 2023.12.x faster.
I will check RAM etc and advise.

v2024.3.6
41.09s - 55.56s - 36.49s - 36.15s - 34.08 

v2023.12.6
27.57s - 20.82s - 20.76s

**Reply by MarcinZiabek** [reply]
Have you tried using the `UseOriginalImage` option?

**Reply by Craigoman79** [reply]
Ok played around with things. I am sure I can find a combination of compression of the images before together with `UseOriginalImage` to get the best optimal performance. I'll try to batch up the generation for large files and then merge them. Thanks for the help.

I'll let you know how it goes.

**Reply by Craigoman79** [reply]
batching works which i then merge at the end. the library performance is good!
similar to another topic. ideally would like to merge them and then setup the footer to correctly represent page numbers etc. It will be a good addition.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Glad to hear the batching-and-merge approach, combined with `UseOriginalImage` to skip the in-memory resize/compress step, gave you good performance for large image-heavy documents.

On the remaining point about page numbers across merged documents: since this thread was opened, the [Document Operations](https://www.questpdf.com/concepts/document-operations.html) API (version 2024.12.0) added native merging, which together with Sections lets you apply continuous, correct page numbering across the combined output — so your batch-then-merge workflow can now carry proper page numbers in the footer.

Should you encounter any further problems, please don't hesitate to open a new issue.
Thank you!
